cli/src/cmd/test.rs

```diff
@@ -311,124 +315,123 @@ fn test<A: ArtifactOutput + 'static>(
 ) -> eyre::Result<TestOutcome> {
     let verbosity = evm_opts.verbosity;
     let gas_reporting = gas_reports.0;
-
     if gas_reporting && evm_opts.verbosity < 3 {
         // force evm to do tracing, but dont hit the verbosity print path
         evm_opts.verbosity = 3;
     }
-
     let mut runner = builder.build(project, evm_opts)?;
 
-    let results = runner.test(&filter)?;
-
-    let mut gas_report = GasReport::new(gas_reports.1);
-
-    let (funcs, events, errors) = runner.execution_info;
     if json {
-        let res = serde_json::to_string(&results)?;
+        let results = runner.test(&filter, None)?;
+        let res = serde_json::to_string(&results)?; // TODO: Make this work normally
         println!("{}", res);
+        Ok(TestOutcome::new(results, allow_failure))
     } else {
         // Dapptools-style printing of test results
-        for (i, (contract_name, tests)) in results.iter().enumerate() {
-            if i > 0 {
-                println!()
-            }
-            if !tests.is_empty() {
-                let term = if tests.len() > 1 { "tests" } else { "test" };
-                println!("Running {} {} for {}", tests.len(), term, contract_name);
-            }
-
-            for (name, result) in tests {
-                // build up gas report
-                if gas_reporting {
-                    if let (Some(traces), Some(identified_contracts)) =
-                        (&result.traces, &result.identified_contracts)
-                    {
-                        gas_report.analyze(traces, identified_contracts);
-                    }
+        let mut gas_report = GasReport::new(gas_reports.1);
+        let execution_info = runner.execution_info.clone();
+        let known_contracts = runner.known_contracts.clone();
+        let index = Arc::new(Mutex::new(0));
+
+        let results = runner.test(
+            &filter,
+            Some(Box::new(move |contract_name: String, tests: BTreeMap<String, TestResult>| {
+                // This mutex syncs stdout per contract, do not remove even if the newline below is
+                let mut index = index.lock().unwrap();
+                if *index > 0 {
+                    println!()
```

Comment:
What do you think of replacing the mutex and stream function with an mpsc channel? Rayon has a [`map_with`](https://docs.rs/rayon/latest/rayon/iter/trait.ParallelIterator.html#method.map_with) that plays really nicely with it, and it should solve the stdout issue the mutex is trying to solve

Reply:
`enumerate()` won't actually work in this case. It would mark the first contract in the list, not the first contract finished. Also, the filtering being done might remove the zeroth item anyway. (Adding it later in the chain would make the chain sync and defeat the purpose.)

The extra newline at the beginning of the test output doesn't matter so I just removed the index check entirely.

Moving to `mpsc` is a lot cleaner imo and we get to axe the `Mutex` so good call 👍